app/jobs/plexer_job.rb

```diff
@@ -52,6 +52,10 @@ def find_or_create_unreplicated_part(zmv, part_s3_key, metadata)
       size: metadata[:size],
       suffix: File.extname(part_s3_key)
     ) { |part| part.unreplicated! }
+  rescue ActiveRecord::RecordNotUnique
+    # This catches an exception triggered by a race condition because find_or_create_by is not atomic.
+
+    retry
```

Comment:
i'm a bit worried this ends up in an infinite loop in the rare case where it might get triggered, but i haven't really used `retry` myself, so i don't have much intuition for it.  but from what i was reading just now, it seems like it would try the same thing over from the top of the `find_or_create_unreplicated_part` method with the same `metadata` (because that seems like the implicit `begin` for the `retry`'s `rescue`?).  and if there's already a part for that ZMV with that suffix, i'd expect the constraint to get violated again, and then the same error/retry.

or,  am i misunderstanding, and what's retried is the next iteration of the `zmvs.each` loop?

i'm a little rusty on it since i haven't really touched it during this maintenance WC, but there are delivery pipeline integration tests in pres cat, so i think we could simulate this scenario, though not sure off-hand how much of a pain it'd be.

Reply:
@jmartin-sul the exception would be raised by `find_or_create_by`, and that method first tries a `find_by` before issuing a `create`. Presumably subsequent retries would not fail because the `find_by` returns a record (it does not raise `ActiveRecord::RecordNotUnique`). So if there is already a part for that ZMV with that suffix, it ought to get returned. Does that assuage your infinite loop concerns? It should only get triggered when the momentary race condition is encountered.

Reply:
ah!  yes.  that makes sense.  thanks!